internal/stream/stream.go

```diff
@@ -0,0 +1,313 @@
+package stream
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/frostbyte73/core"
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/livekit/psrpc"
+	"github.com/livekit/psrpc/internal"
+	"github.com/livekit/psrpc/internal/bus"
+	"github.com/livekit/psrpc/internal/interceptors"
+	"github.com/livekit/psrpc/internal/logger"
+	"github.com/livekit/psrpc/internal/rand"
+	"github.com/livekit/psrpc/pkg/info"
+)
+
+type Stream[SendType, RecvType proto.Message] interface {
+	psrpc.ServerStream[SendType, RecvType]
+
+	Ack(context.Context, *internal.Stream) error
+	HandleStream(is *internal.Stream) error
+	Hijacked() bool
+}
+
+type StreamAdapter interface {
+	Send(ctx context.Context, msg *internal.Stream) error
+	Close(streamID string)
+}
+
+type stream[SendType, RecvType proto.Message] struct {
+	*streamBase[SendType, RecvType]
+
+	handler  psrpc.StreamHandler
+	hijacked bool
+}
+
+type streamBase[SendType, RecvType proto.Message] struct {
+	psrpc.StreamOpts
+
+	ctx      context.Context
+	cancel   context.CancelFunc
+	streamID string
+
+	adapter  StreamAdapter
+	recvChan chan RecvType
+
+	mu      sync.Mutex
+	pending atomic.Int32
+	acks    map[string]chan struct{}
+	closed  core.Fuse
+	err     error
+}
+
+func NewStream[SendType, RecvType proto.Message](
+	ctx context.Context,
+	i *info.RequestInfo,
+	streamID string,
+	timeout time.Duration,
+	adapter StreamAdapter,
+	streamInterceptors []psrpc.StreamInterceptor,
+	recvChan chan RecvType,
+	acks map[string]chan struct{},
+) Stream[SendType, RecvType] {
+
+	ctx, cancel := context.WithCancel(ctx)
+	base := &streamBase[SendType, RecvType]{
+		StreamOpts: psrpc.StreamOpts{Timeout: timeout},
+		ctx:        ctx,
+		cancel:     cancel,
+		streamID:   streamID,
+		adapter:    adapter,
+		recvChan:   recvChan,
+		acks:       acks,
+		closed:     core.NewFuse(),
+	}
+
+	return &stream[SendType, RecvType]{
+		streamBase: base,
+		handler: interceptors.ChainClientInterceptors[psrpc.StreamHandler](
+			streamInterceptors, i, base,
+		),
+	}
+}
+
+func (s *stream[SendType, RecvType]) HandleStream(is *internal.Stream) error {
+	if s.closed.IsBroken() {
+		return psrpc.ErrStreamClosed
+	}
```

Comment:
this should be in the message switch case. we need to continue processing acks after local close otherwise the pending sends can only be resolved by timing out